packages/babylon/src/plugins/typescript.js

```diff
@@ -2084,4 +2087,8 @@ export default (superClass: Class<Parser>): Class<Parser> =>
     shouldParseAsyncArrow(): boolean {
       return this.match(tt.colon) || super.shouldParseAsyncArrow();
     }
+
+    canHaveLeadingDecorator() {
+      return this.isAbstractClass() || super.canHaveLeadingDecorator();
```

Comment:
Can you swap these conditions, so that we avoid the unnecessary lookahead in most cases?